src/Chainweb/Pact/SPV.hs

```diff
@@ -169,7 +177,77 @@ verifySPV bdb bh typ proof = runExceptT $ go typ proof
+--   of the legacy bytestring parser; and improperly padded messages
+--   will get extra padding, because that is the legacy chainweb behavior.
+--
+--   `Simplified` errors are a static string, which may not describe
+--   the issue with the base64-encoded string as well, but will be
+--   more stable as we upgrade base64 decoding libraries in the future.
+--   In the `Simplified` errors setting, messages rejected for using
+--   non-canonical encodings will remain as errors, because we want to
+--   begin enforcing the expected constraints on base64-encoded messages.
+--   Similarly, Simplified parsing will not add extra padding to improperly
+--   padded messages.
+data GenerateBase64ErrorMessage
+  = Legacy
+  | Simplified
+
+
+-- | A modified version of `decodeBase64UrlNoPaddingText` that emits
+--   base64 decoding errors in a configurable way.
+decodeB64UrlNoPaddingTextWithFixedErrorMessage :: MonadThrow m => GenerateBase64ErrorMessage -> T.Text -> m B.ByteString
+decodeB64UrlNoPaddingTextWithFixedErrorMessage errorMessageType msg =
+  fromEitherM
+  . first (\e -> Base64DecodeException $ base64ErrorMessage e)
+  . (if errorMessageType == Legacy then patchNonCanonical else id)
+  . B64U.decode
+  . T.encodeUtf8
+  . (if errorMessageType == Legacy then pad else id)
+  $ msg
+  where
+    pad t = let s = T.length t `mod` 4 in t <> T.replicate ((4 - s) `mod` 4) "="
+    base64ErrorMessage m = case errorMessageType of
+      Legacy -> base64DowngradeErrorMessage (Text.pack m)
+      Simplified -> "could not base64-decode message"
+    patchNonCanonical decodeResult = case decodeResult of
+      Right bs -> Right bs
+      Left e | "non-canonical" `T.isInfixOf` T.pack e ->
+               Right (B64U.decodeLenient (T.encodeUtf8 msg))
+      Left e -> Left e
+{-# INLINE decodeB64UrlNoPaddingTextWithFixedErrorMessage #-}
+
+
+
+-- | Converts the error message format of base64-bytestring-1.2
+--   into that of base64-bytestring-0.1, for the error messages
+--   that have made it onto the chain.
+--   This allows us to upgrade to base64-bytestring-1.2 without
+--   breaking compatibility.
+base64DowngradeErrorMessage :: Text -> Text
+base64DowngradeErrorMessage msg = case msg of
+  "Base64-encoded bytestring has invalid size" ->
+       "invalid base64 encoding near offset 0"
+  (Text.stripPrefix "invalid character at offset: " -> Just suffix) ->
+    Text.pack $ "invalid base64 encoding near offset " ++ show (adjustedOffset suffix)
+  (Text.stripPrefix "invalid padding at offset: " -> Just suffix) ->
+    Text.pack $ "invalid padding near offset " ++ show (adjustedOffset suffix)
+  e -> e
+  where
+    adjustedOffset :: Text -> Int
+    adjustedOffset suffix = case readMaybe (Text.unpack suffix) of
+      Nothing -> 0
+      Just offset -> let
+        endsWithThreeEquals = T.drop (T.length msg - 3) msg == "==="
+        adjustment = if endsWithThreeEquals then -1 else 0
+        in
+        offset - (offset `rem` 4) + adjustment
```

Comment:
These calculations patch the difference between the old decode messages generated from improper padding (e.g. `X===`) and the new ones which fail immediately.